# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Verifier.cpp
# Commit: f7b553bd9c0b
# Full Hash: f7b553bd9c0b77b0130d377f9a9c696c8c611a01
# Author: Steve Fink <sfink@mozilla.com>
# Date: 2021-05-11 21:39:06
# Regressor Bug: 1694538
# File Overlap Count: 1
# Description:
#   Bug 1694538 - Convert gcWeakKeys from <weakmap, key> tuples that require an extra lookup, to <ptr to color, target> tuples that simplify the code substantially r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D106226
# ==============================================================================

diff -r 89f7028049ed -r f7b553bd9c0b js/src/gc/Verifier.cpp
--- a/js/src/gc/Verifier.cpp	Tue May 11 14:01:20 2021 +0000
+++ b/js/src/gc/Verifier.cpp	Tue May 11 14:18:38 2021 +0000
@@ -539,9 +539,9 @@
    * For saving, smush all of the keys into one big table and split them back
    * up into per-zone tables when restoring.
    */
-  gc::WeakKeyTable savedWeakKeys(SystemAllocPolicy(),
-                                 runtime->randomHashCodeScrambler());
-  if (!savedWeakKeys.init()) {
+  gc::EphemeronEdgeTable savedEphemeronEdges(
+      SystemAllocPolicy(), runtime->randomHashCodeScrambler());
+  if (!savedEphemeronEdges.init()) {
     return;
   }
 
@@ -551,15 +551,15 @@
     }
 
     AutoEnterOOMUnsafeRegion oomUnsafe;
-    for (gc::WeakKeyTable::Range r = zone->gcWeakKeys().all(); !r.empty();
-         r.popFront()) {
+    for (gc::EphemeronEdgeTable::Range r = zone->gcEphemeronEdges().all();
+         !r.empty(); r.popFront()) {
       MOZ_ASSERT(r.front().key->asTenured().zone() == zone);
-      if (!savedWeakKeys.put(r.front().key, std::move(r.front().value))) {
+      if (!savedEphemeronEdges.put(r.front().key, std::move(r.front().value))) {
         oomUnsafe.crash("saving weak keys table for validator");
       }
     }
 
-    if (!zone->gcWeakKeys().clear()) {
+    if (!zone->gcEphemeronEdges().clear()) {
       oomUnsafe.crash("clearing weak keys table for validator");
     }
   }
@@ -650,18 +650,19 @@
   for (GCZonesIter zone(gc); !zone.done(); zone.next()) {
     WeakMapBase::unmarkZone(zone);
     AutoEnterOOMUnsafeRegion oomUnsafe;
-    if (!zone->gcWeakKeys().clear()) {
+    if (!zone->gcEphemeronEdges().clear()) {
       oomUnsafe.crash("clearing weak keys table for validator");
     }
   }
 
   WeakMapBase::restoreMarkedWeakMaps(markedWeakMaps);
 
-  for (gc::WeakKeyTable::Range r = savedWeakKeys.all(); !r.empty();
+  for (gc::EphemeronEdgeTable::Range r = savedEphemeronEdges.all(); !r.empty();
        r.popFront()) {
     AutoEnterOOMUnsafeRegion oomUnsafe;
     Zone* zone = r.front().key->asTenured().zone();
-    if (!zone->gcWeakKeys().put(r.front().key, std::move(r.front().value))) {
+    if (!zone->gcEphemeronEdges().put(r.front().key,
+                                      std::move(r.front().value))) {
       oomUnsafe.crash("restoring weak keys table for validator");
     }
   }